orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,292 @@
+package cluster
+
+import (
+	"sync"
+	"time"
+
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric/bccsp"
+	"github.com/hyperledger/fabric/common/flogging"
+	"github.com/pkg/errors"
+)
+
+type suspectSet map[string]struct{}
+
+func (s suspectSet) insert(entry string) {
+	s[entry] = struct{}{}
+}
+
+func (s suspectSet) remove(entry string) {
+	delete(s, entry)
+}
+
+func (s suspectSet) has(entry string) bool {
+	_, ok := s[entry]
+	return ok
+}
+
+//go:generate mockery -dir . -name BlockSource -case underscore -output mocks/
+
+type BlockSource interface {
+	PullBlock(seq uint64) *common.Block
+	UpdateEndpoints(endpoints []EndpointCriteria)
+	Close()
+}
+
+type FetcherConfig struct {
+	// TODO: The fields of this struct are evolving
+	Endpoints    []EndpointCriteria
+	FetchTimeout time.Duration
+}
+
+type BlockFetcher struct {
+	MaxPullBlockRetries      uint64
+	BlockSourceFactory       func(c FetcherConfig) BlockSource
+	Config                   FetcherConfig
+	Endpoints                []EndpointCriteria
+	suspects                 suspectSet // a set of bft suspected nodes.
+	Logger                   *flogging.FabricLogger
+	currentEndpoint          EndpointCriteria
+	currentSeq               uint64
+	ShuffleTimeout           time.Duration
+	LastShuffledAt           time.Time
+	MaxByzantineFaults       int64
+	ConfirmByzantineBehavior func(attestations []*common.Block) bool
+}
+
+func (bf *BlockFetcher) pullBlockFromSource(seq uint64) *common.Block {
+	bf.currentSeq = seq
+
+	// create a local config
+	config := bf.Config
+	// fill only current endpoint criteria
+	config.Endpoints = []EndpointCriteria{bf.currentEndpoint}
+
+	blockSource := bf.BlockSourceFactory(config)
+	// TODO: Check if the below line is redundant
+	blockSource.UpdateEndpoints([]EndpointCriteria{bf.currentEndpoint})
+	// pull block
+	return blockSource.PullBlock(seq)
+}
+
+func (bf *BlockFetcher) setSomeEndpoint() {
+	// Connect to first endpoint for the first time
+	// in case of shuffletimeout expires, shuffle the endpoint
+	// otherwise continue with the last endpoint
+
+	// choose endpoint if there is none
+	emptyEC := EndpointCriteria{}
+	if bf.currentEndpoint.Endpoint == emptyEC.Endpoint {
+		// first time allocation
+		bf.Logger.Info("Set First Endpoint")
+		bf.currentEndpoint = bf.Endpoints[0]
+		bf.LastShuffledAt = time.Now()
```

Comment:
instead of using real time, you can inject a function that returns a time, and then you can make the tests time independent. 

```
type TimeFunc func() time.Time
```

Reply:
agree